modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -0,0 +1,260 @@
+            selector_address=f"{self.modyn_config['selector']['hostname']}:{self.modyn_config['selector']['port']}",
+            num_prefetched_partitions=num_prefetched_partitions,
+            parallel_prefetch_requests=parallel_prefetch_requests,
+            tokenizer=tokenizer,
+        )
+
+    def _init_model_downloader(self) -> None:
+        assert self.pipeline_config is not None
+        assert self.modyn_config is not None
+
+        self.model_downloader = ModelDownloader(
+            self.modyn_config,
+            self.pipeline_id,
+            self.pipeline_config["training"]["device"],
+            self.base_dir,
+            f"{self.modyn_config['model_storage']['hostname']}:{self.modyn_config['model_storage']['port']}",
+        )
+
+    def _create_dirs(self) -> None:
+        assert self.pipeline_id is not None
+        assert self.base_dir is not None
+
+        self.exp_output_dir = self.base_dir / str(self.pipeline_id) / "datadrift"
+        self.drift_dir = self.exp_output_dir / "drift_results"
+        os.makedirs(self.drift_dir, exist_ok=True)
+
+    def init_trigger(self, pipeline_id: int, pipeline_config: dict, modyn_config: dict, base_dir: pathlib.Path) -> None:
+        self.pipeline_id = pipeline_id
+        self.pipeline_config = pipeline_config
+        self.modyn_config = modyn_config
+        self.base_dir = base_dir
+
+        if len(self.trigger_config) > 0:
+            self._parse_trigger_config()
+
+        self._init_dataloader_info()
+        self._init_model_downloader()
+        self._create_dirs()
+
+    def run_detection(self, reference_embeddings_df: pd.DataFrame, current_embeddings_df: pd.DataFrame) -> bool:
+        assert self.dataloader_info is not None
+        assert self.drift_dir is not None
+
+        # Run Evidently detection
+        column_mapping = ColumnMapping(embeddings={self.evidently_column_mapping_name: reference_embeddings_df.columns})
+
+        # https://docs.evidentlyai.com/user-guide/customization/embeddings-drift-parameters
+        report = Report(metrics=self.metrics)
+        report.run(
+            reference_data=reference_embeddings_df, current_data=current_embeddings_df, column_mapping=column_mapping
+        )
+        result = report.as_dict()
+        result_print = [
+            (x["result"]["drift_score"], x["result"]["method_name"], x["result"]["drift_detected"])
+            for x in result["metrics"]
+        ]
+        logger.info(
+            f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}][Dataset {self.dataloader_info.dataset_id}]"
+            + f"[Result] {result_print}"
+        )
+
+        with open(self.drift_dir / f"drift_{self.previous_trigger_id}.json", "w", encoding="utf-8") as f:
+            json.dump(result, f)
+
+        return result["metrics"][0]["result"]["drift_detected"]
+
+    def detect_drift(self, idx_start: int, idx_end: int) -> bool:
```

Comment:
Can you also add a docstring here with some more context (e.g., we always compare against the previous trigger in the current implementation, right?)?

Reply:
done